text/0000-default-field-values.md

```diff
@@ -0,0 +1,1807 @@
+
+```rust
+fn open_window(config: { height: u32, width: u32 }) {
+    // logic...
+}
+
+open_window({ height: 720, width: 1280 });
+```
+
+Since this proposal introduces field defaults, the natural combination with
+structural records would be to permit them to have defaults. For example:
+
+```rust
+fn open_window(config: { height: u32 = 1080, width: u32 = 1920 }) {
+    // logic...
+}
+```
+
+A coercion could then allow you to write:
+
+```rust
+open_window({ .. });
+```
+
+This could be interpreted as `open_window({ RangeFull })`, see the previous
+section for a discussion... alternatively `open_window(_)` could be permitted
+instead for general value inference where `_` is a placeholder expression
+similar to `_` as a type expression placeholder
+(i.e. a fresh and unconstrained unification variable).
+
+If you wanted to override a default, you would write:
+
+```rust
+open_window({ height: 720, });
+```
+
+Note that the syntax used to give fields in structural records defaults belongs
+to the type grammar; in other words, the following would be legal:
+
+```rust
+type RGB = { red: u8 = 0, green: u8 = 0, blue: u8 = 0 };
+
+let color: RGB = { red: 255, };
+```
+
+As structural records are not yet in the language,
+figuring out designs for how to extend this RFC to them is left
+as possible work for the future.
+
+
+## Integration with struct literal type inference
+
+Yet another common requested feature is the introduction of struct literal type
+inference in the form of elision of the name of an ADT literal when it can be
+gleaned from context. This has sometimes been proposed as an alternative or
+complementary to structural records. This would allow people to write
+`foo(_ { bar: 42 })` where the function argument type is inferred from the `foo`
+definition. struct literal type inference with default struct fields would also
+allow people to write APIs that "feel" like named function arguments when
+calling them, although not when defining them.
+
+```rust
+struct Config {
+    height: u32 = 1080,
+    width: u32 = 1920,
+}
+fn open_window(config: Config) {
+    // logic...
+}
+
+open_window(_ { width: 800, .. });
+```
+
+## Accessing default values from the type
+
+If one were to conceptualize default field values in the following way:
+
+```rust
+struct Config {
+    height: u32 = Self::HEIGHT,
+    width: u32 = Self::WIDTH,
+}
+
+impl Config {
+    const HEIGHT: u32 = 1080,
+    const WIDTH: u32 = 1920,
+}
+```
+
+It would follow that one should be able to access the value of these defaults
+*without* constructing `Config`, by writing `Config::HEIGHT`. I do not believe
+this should be done or advanced, but there's nothing in this RFC that precludes
+*some* mechanism to access these values in the future. With the RFC as written,
+these values can be accessed by instantiating `Config { .. }.height`, as long
+as `height` is visible in the current scope.
+
+## Non-const values
```

Comment:
If we did this we could immediately start allowing use of this syntax for all `Default` impls.

Related thought.. Should there be a way of opting into - or out of - a guarantee of const init? What would syntax for that look like? Perhaps this?

```rust
struct Foo const {
  height: u32 = 0,
  width: u32 = 0,
}
```

Reply:
> Related thought.. Should there be a way of opting into - or out of - a guarantee of const init? What would syntax for that look like?

Wont existing const blocks be enough?
At least for const init.

```rust
struct Foo {
  height: u32 = const { 0 },
  width: u32 = const { 0 },
}
```

Maybe in the future syntax sugar can be added.

Reply:
I guess it may be useful for heap allocations, for example for owned collections? I don't think that is something that will ever be solved with const, proposals talk about having a separate const heap, but there is no need to have separate heap here.
Not sure whether really prohibiting allocations here is a good thing or not, there are a lot of places in Rust where you can have stupid implicit behaviour and launch missiles, for example during deref. I personally don't really like allocations and control flow not being obvious, but Rust is already bad at it, this is not Zig.

I think it is ok to just run with const only for now, then we can test this feature, and see how bad not having side effects actually is.

Reply:
> A different approach which I'm exploring here is to make the `..` syntax not be always-const-friendly.

I wonder if it maybe doesn't make sense to allow the defaults to be non-const, *but always* warn-by-default lint when that feature is leveraged. This means that someone that really needs can use it and has to explicitly opt-into it by marking the crate, module, item, field with an `#[allow(non_const_defaults)]`. This pushes back against egregious accidental usage of the feature, while still allowing it for the cases where it is needed for corner-case situations. Note that the intent for this RFC is to allow for something that `derive(Default)` doesn't let you do: types where *some* fields are mandatory. It is not an attempt of making `S { ..Default::default() }` less verbose.

I also believe that starting with `const` expressions first and extend support for non-`const` expressions later will make the process of implementing and stabilizing each independent part much smoother, while not crossing any one-way doors and providing a lot of value earlier rather than later.